File: single/echo-server.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/socket.h>

#include "echo.h"

int server() {
  int listen_socket = listen_to_echo_port();
  int sock = accept_client(listen_socket);

  for(;;) {
    char* buf = malloc(8);
    ssize_t read = recv(sock, buf, 8, 0);
    if(read == 0) {
      free(buf);
      break;
    }
    send(sock, buf, 8, 0);
    free(buf);
  }

  return 0;
}

int
main(int argc, char** argv) {
  printf("Echo server started.\n");
  server();
  printf("Echo server finished.\n");
  return 0;
}

File: single/echo-server.hs
{-# LANGUAGE CPP, BangPatterns #-}

#include "echodef.h"

module Main where

import Control.Concurrent
import Control.Monad
import qualified Data.ByteString as BS
import Network.Socket
import qualified Network.Socket.ByteString as NBS

echo :: Socket -> IO ()
echo sock = do
  bs <- NBS.recv sock ECHO_SIZE
  when (BS.length bs > 0) $ do
    NBS.send sock bs
    echo sock

server :: IO ()
server = do
    (addr:_) <- getAddrInfo (Just (defaultHints { addrFlags = [AI_PASSIVE] }))
                            Nothing
                            (Just ECHO_PORT)
    listenSock <- socket (addrFamily addr) Stream defaultProtocol
    setSocketOption listenSock ReuseAddr 1
    bindSocket listenSock (addrAddress addr)

    listen listenSock MAX_BACKLOG
    (sock, _) <- accept listenSock
    setSocketOption sock NoDelay 1

    echo sock

main :: IO ()
main = do
    putStrLn "Echo server started."
    runInUnboundThread server
    putStrLn "Echo server finished."
